File: core/src/test/kotlin/io/gpxtoign/core/layout/Tracks.kt
package io.gpxtoign.core.layout

import io.gpxtoign.core.geo.L93
import io.gpxtoign.core.geo.Lambert93
import io.gpxtoign.core.gpx.GpxFile
import io.gpxtoign.core.gpx.GpxSegment

/** Builds a synthetic GPX from Lambert-93 metres so tests can reason in ground distance. */
object Tracks {
    private const val BASE_X = 900_000.0
    private const val BASE_Y = 6_400_000.0

    fun of(vararg offsets: Pair<Double, Double>): GpxFile = of(offsets.toList())

    fun of(offsets: List<Pair<Double, Double>>): GpxFile = GpxFile(
        name = "synthetic.gpx",
        segments = listOf(
            GpxSegment(
                name = null,
                points = offsets.map { (dx, dy) ->
                    Lambert93.inverse(L93(BASE_X + dx, BASE_Y + dy))
                },
            ),
        ),
    )

    fun line(dx: Double, dy: Double, steps: Int = 200): GpxFile =
        of((0..steps).map { i -> dx * i / steps to dy * i / steps })
}

File: packages/core/test/layout/tracks.ts
import { inverse } from '../../src/geo/lambert93.js'
import type { GpxFile } from '../../src/gpx/gpx.js'

const BASE_X = 900_000.0
const BASE_Y = 6_400_000.0

/** Builds a synthetic GPX from Lambert-93 metres so tests can reason in ground distance. */
export function track(offsets: readonly (readonly [number, number])[]): GpxFile {
  return {
    name: 'synthetic.gpx',
    segments: [
      {
        name: null,
        points: offsets.map(([dx, dy]) => inverse(BASE_X + dx, BASE_Y + dy)),
      },
    ],
  }
}

export function line(dx: number, dy: number, steps = 200): GpxFile {
  return track([...Array(steps + 1)].map((_, i): [number, number] => [(dx * i) / steps, (dy * i) / steps]))
}
